Public/Src/Engine/Processes/SharedOpaqueOutputHelper.cs

```diff
@@ -180,19 +180,56 @@ private static bool IsSharedOpaqueOutputWithFallback(string expandedPath, bool c
             }
         }
 
+        private const int MaxNumberOfAttemptsForMarkingSharedOpaqueOutputs = 3;
+        private static readonly TimeSpan SleepDurationBetweenMarkingAttempts = TimeSpan.FromMilliseconds(100);
+
         /// <summary>
         /// Marks a given path as "shared opaque output"
         /// </summary>
+        /// <remarks>
+        /// Retries are needed because: (Win|Unix).SetPathAsSharedOpaqueOutput does something like
+        ///   - check if file has write access rights
+        ///   - if it doesn't, give it those rights
+        ///   - mark the file as shared opaque output
+        ///   - ...
+        /// There is a race between checking and setting access rights, so it is possible that

```

Comment:
I'm not denying that this is happening, and retries are always good :). But I'm curious about what's really going on. From my understanding the cache cannot be racing with this (the pip ran, so we actually set the marker before pushing this to the cache, or it is a cache hit and then we change the marker during materialization - if the marker is not there yet). And any race between pips mean a double write anyway.

Reply:
All good questions.  I don't know what exactly happened.  In one build I saw, a file was just produced by a COPYFILE pip, and then marking it subsequently failed with EACCESS.  So it is possible that retrying will not help in this case, but right now that's all we can do.